File: backend/app/schemas/routine.py
from pydantic import BaseModel, ConfigDict, Field


class RoutineExerciseBase(BaseModel):
    exercise_id: int
    target_sets: int = Field(gt=0)
    order: int = Field(ge=0)
    # Nulo = ejercicio suelto. Mismo entero = mismo bloque de super-serie dentro
    # de esta rutina (solo comparable dentro de la misma rutina). Validado en el
    # router: un valor usado por menos de 2 exercise_id distintos es un 400.
    superset_group: int | None = None


class RoutineExerciseCreate(RoutineExerciseBase):
    pass


class RoutineExercise(RoutineExerciseBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    routine_id: int


class RoutineBase(BaseModel):
    name: str


class RoutineCreate(RoutineBase):
    exercises: list[RoutineExerciseCreate] = []


class Routine(RoutineBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    exercises: list[RoutineExercise] = []

File: mobile/src/types/routine.ts
// superset_group: null = ejercicio suelto; mismo entero = mismo bloque de
// super-serie. El ámbito de comparación es solo dentro de esta rutina, no es
// un id global (dos rutinas distintas pueden reusar el mismo número).
export type RoutineExercise = {
  id: number;
  routine_id: number;
  exercise_id: number;
  target_sets: number;
  order: number;
  superset_group: number | null;
};

export type RoutineExerciseCreate = {
  exercise_id: number;
  target_sets: number;
  order: number;
  superset_group: number | null;
};

export type Routine = {
  id: number;
  name: string;
  exercises: RoutineExercise[];
};

export type RoutineCreate = {
  name: string;
  exercises: RoutineExerciseCreate[];
};
